src/Google.Api.Gax/PagedAsyncEnumerable.cs

```diff
@@ -0,0 +1,86 @@
+﻿/*
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+
+using Google.Protobuf;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Api.Gax
+{
+    public sealed class PagedAsyncEnumerable<TRequest, TResponse, TResource> : IPagedAsyncEnumerable<TResponse, TResource>
+        where TRequest : class, IPageRequest, IMessage<TRequest>
+        where TResponse : class, IPageResponse<TResource>, IMessage<TResponse>
+    {
+        private readonly CallSettings _callSettings;
+        private readonly TRequest _request;
+        private readonly ApiCall<TRequest, TResponse> _apiCall;
+
+        public PagedAsyncEnumerable(CallSettings callSettings,
+            TRequest request, ApiCall<TRequest, TResponse> apiCall)
+        {
+            _callSettings = callSettings;
+            _request = request;
+            _apiCall = apiCall;
+        }
+
+        public IPagedAsyncEnumerator<TResponse> GetEnumerator() =>
+            new PagedAsyncEnumerator(_callSettings, _request.Clone(), _apiCall);
+
+        IAsyncEnumerator<TResponse> IAsyncEnumerable<TResponse>.GetEnumerator() => GetEnumerator();
+
+        private class PagedAsyncEnumerator : IPagedAsyncEnumerator<TResponse>
+        {
+            private readonly CallSettings _callSettings;
+            private readonly ApiCall<TRequest, TResponse> _apiCall;
+            private readonly TRequest _request; // This is mutated during iteration
+            private bool _finished;
+
+            public PagedAsyncEnumerator(CallSettings callSettings,
+                TRequest request, ApiCall<TRequest, TResponse> apiCall)
+            {
+                _callSettings = callSettings;
+                _request = request;
+                _apiCall = apiCall;
+            }
+
+            public TResponse Current { get; private set; }
+
+            public async Task<bool> MoveNext(CancellationToken cancellationToken)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                if (_finished)
+                {
+                    return false;
+                }
+                CallSettings effectiveCallSettings = _callSettings;
+                if (cancellationToken != default(CancellationToken))
+                {
+                    effectiveCallSettings = _callSettings?.Clone() ?? new CallSettings();
```

Comment:
If we had a CallSettings(CallSettings) constructor, this would be simpler:

```csharp
effectiveCallSettings = new CallSettings(_callSettings) { CancellationToken = cancellationToken };
```